### PR TITLE
Enable diffing m2m fields

### PR DESCRIPTION
**Pull request by thijskramer**
At first [@avalanchy](https://github.com/avalanchy) did the initial work. After that https://github.com/bmedx made an excellent PR based on his work which wasn't completed, so I tried to make it work and created a PR myself, based on https://github.com/edx-unsupported/django-simple-history/pull/1 and rebased on the current master. 

## Description

This PR enables tracking history of m2m relations. It works in my specific situation, however I  would be happy to hear if there are any situations I haven't looked at, I should cover or simply could be improved.

## Related Issue
#399 - Track history of M2M fields

## Motivation and Context
A few days ago I received the question "how can I see who has removed an article from this category?" I immediately dove into the history I had enabled, but I quickly discovered that  m2m relations weren't tracked. I started googling for solutions and stumbled upon the Pull Request @bmedx made for #399. 
I installed his fork and after some minor tweaks and a rebase, I made it work for my situation. 
It tracks each m2m change in a separate m2m_history model, so that you won't need a reference to the actual referenced model.

## How Has This Been Tested?
Tests have been included, however I'm unsure if they are enough to cover all use cases.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have run the `pre-commit run` command to format and lint.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] I have added my name and/or github handle to `AUTHORS.rst`
- [X] I have added my change to `CHANGES.rst`
- [X] All new and existing tests passed.

### CODE REVIEW

File: simple_history/models.py [hunk 1]
```diff
@@ -655,6 +755,35 @@ def diff_against(self, old_history, excluded_fields=None, included_fields=None):
...
+            history_field_names = []
+            if reference_history_m2m_item:
+                # Create a list of field names to compare against.
+                # The list is generated without the primary key of the intermediate
+                # table, the foreign key to the history record, and the actual 'history'
+                # field, to avoid false positives while diffing.
+                history_field_names = [
+                    f.name
+                    for f in reference_history_m2m_item._meta.fields
+                    if f.name not in ["id", "m2m_history_id", "history"]
```
Comment (thijskramer) [inline simple_history/models.py:774]: Would it be more sensible to apply the same logic here as in #923 ?

Comment (jeking3) [reply simple_history/models.py:774]: If you mean checking for the editable flag, yes, I think that makes sense.

https://github.com/jazzband/django-simple-history/pull/923/files#diff-a13c48090526a77430e6819c61ffb009b8ed15ffca1b40f3eef675c8c9b77db9R640

Comment (thijskramer) [reply simple_history/models.py]: I've added that and rebased on master!

**Comment by jeking3** [comment]
I wonder if perhaps the PR title is not quite right, since this seems to be doing more than allowing for a diff.

**Comment by lazarust** [comment]
@thijskramer It looks like some tests failed in the GH Action, could you look into that? This feature would be really helpful for me!

**Comment by thijskramer** [comment]
I'll take a look in the upcoming days!

**Comment by utapyngo** [comment]
@lazarust can you work this around by converting your m2m relation into an explicit model and adding history to it?
I really want #866 to be released ASAP, and this PR has conflicts and broken tests.

**Review by lazarust — DISMISSED** [review]
I've looked through this and it looks good to me! Hopefully it can get merged in soon!!

**Comment by thijskramer** [comment]
I have to fix a few tests before it can be merged, but I’m a bit short on
time so I’d welcome any help!

**Comment by legau** [comment]
I fixed your tests on https://github.com/thijskramer/django-simple-history/pull/20
This should be good once merged into your branch

**Comment by legau** [comment]
One more to go https://github.com/thijskramer/django-simple-history/pull/21 to resolve conflicts with master

**Comment by thijskramer** [comment]
> 

I think it's cleaner to rebase this branch onto `master`, isn't it?

**Comment by thijskramer** [comment]
> Indeed I tried that but with my own branch so it conflicts with your branch now

I've rebased my branch!

**Comment by legau** [comment]
I added inheritance support with https://github.com/thijskramer/django-simple-history/pull/22

**Comment by ipeterov** [comment]
Sorry about a comment that isn't immediately relevant, I just wanted to say that I really want the feature and I'm very happy to see the current activity.

Thank you @thijskramer and @legau for your work!

**Comment by cpackard** [comment]
hey all, any reason this couldn't be merged? m2m field tracking would be a huge plus!

**Review by valberg — APPROVED** [review]

**Comment by cpackard** [comment]
@thijskramer @valberg @legau @johanneswilm thanks for all the work and merging this in 🙌

**Comment by chrisbarber** [comment]
Has this been tested in the django admin? For me with the multi-picker widget it does not reflect the state correctly for historical records. It still just renders the current state.

**Comment by ddabble** [comment]
@chrisbarber Yeah, it's a known bug; see #1063

**Pull request merged**